cluster-controller/src/main/java/io/strimzi/controller/cluster/operations/CreateOperation.java

```diff
@@ -0,0 +1,174 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.strimzi.controller.cluster.operations;
+
+import io.fabric8.kubernetes.api.model.ConfigMap;
+import io.fabric8.kubernetes.api.model.HasMetadata;
+import io.fabric8.kubernetes.api.model.Service;
+import io.fabric8.kubernetes.api.model.extensions.Deployment;
+import io.fabric8.kubernetes.api.model.extensions.StatefulSet;
+import io.fabric8.openshift.api.model.BuildConfig;
+import io.fabric8.openshift.api.model.ImageStream;
+import io.strimzi.controller.cluster.K8SUtils;
+import io.strimzi.controller.cluster.OpenShiftUtils;
+import io.vertx.core.AsyncResult;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.core.Vertx;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class CreateOperation<U, R extends HasMetadata> implements Operation<U> {
```

Comment:
The `CreateOperation` class has now nothing to do with the `CreateXXXClusterOperation` classes. Yet it is on the same level. That seems to be a bit confusing. Moreover the `CreateOperation` has Kubernetes hardcoded in it. I think we should maybe name it differently to make it clear that these are Kubernetes / OpenShift operations.

Reply:
> Moreover the `CreateOperation` has Kubernetes hardcoded in it.

No, the most recent commit abstracted the `*Utils` class into a type parameter `U`, so the same `CreateOperation` can be used as the superclass for both Kube and OpenShift operations.

Reply:
> The `CreateOperation` class has now nothing to do with the `CreateXXXClusterOperation` classes. Yet it is on the same level.

I'm beginning to think that it would make more sense to package the classes differently. Rather than

* `...operations.kubernetes` for operations on Kubernetes resources
* `...operations.openshift` for operations on OpenShift resources
* `...operations` for all the cluster-level operations

maybe it makes more sense to have:

* `...operations.resource` for all (K8s and OS) resource operations
* `...operations.cluster` for all the higher-level cluster operations.

There is also a lot of scope for apply the template method pattern to the `(Create|Delete)(Kafka|Zookeeper|KafkaConnect)ClusterOperation`, since they all follow the same basic pattern of getting an `AbstractCluster` instance, starting a load of async resource operations and then `join()`ing those. A benefit of doing this is it would encapsulate the shared lock handling in one class. 

The update cluster operations are similar, but a bit more complicated.